aarch64/src/interrupts.S

```diff
@@ -0,0 +1,170 @@
+.section .text
+
+.equ	SYNC_INVALID_EL1t,	0
+.equ	IRQ_INVALID_EL1t,	1
+.equ	FIQ_INVALID_EL1t,	2
+.equ	ERROR_INVALID_EL1t,	3
+
+.equ	SYNC_INVALID_EL1h,	4
+.equ	IRQ_INVALID_EL1h,	5
+.equ	FIQ_INVALID_EL1h,	6
+.equ	ERROR_INVALID_EL1h,	7
+
+.equ	SYNC_INVALID_EL0_64,	8
+.equ	IRQ_INVALID_EL0_64,	9
+.equ	FIQ_INVALID_EL0_64,	10
+.equ	ERROR_INVALID_EL0_64,	11
+
+.equ	SYNC_INVALID_EL0_32,	12
+.equ	IRQ_INVALID_EL0_32,	13
+.equ	FIQ_INVALID_EL0_32,	14
+.equ	ERROR_INVALID_EL0_32,	15
+
+.macro	ventry	label
+.align	7
+	b	\label
+.endm
+
+// Aarch64 calling convention dictates that we save:
+// - x0-18
+// - x29 (Frame pointer)
+// - x30 (Link register)
+// We additionally save the registers below, so that they're avaialble via the InterruptFrame
+// - ESR_EL1 (Exception syndrome register EL1)
+// - ELR_EL1 (Exception link register EL1)
+// - FAR_EL1 (Fault address register EL1)
+.macro handle_interrupt type
```

Comment:
Using a macro here is going to duplicate all of this code, which may be a bummer. Often what one does instead is have this in a common place in the code, then for each stub, push a vector number or something and then jump to the common save/restore code.

Reply:
fair point

Reply:
Not sure we can do that easily with aarch64, since we can't write the vector num to the stack without using a register.  In which case you'd end up having some messy interleaving.  Happy to revisit at a later date if there's a nice way to do this.